Core/src/Propagator/detail/CovarianceEngine.cpp

```diff
@@ -44,41 +47,34 @@ FreeToBoundMatrix freeToCurvilinearJacobian(const Vector3& direction) {
   const double x = direction(0);  // == cos(phi) * sin(theta)
   const double y = direction(1);  // == sin(phi) * sin(theta)
   const double z = direction(2);  // == cos(theta)
-  // can be turned into cosine/sine
-  const double cosTheta = z;
-  const double sinTheta = std::hypot(x, y);
-  const double invSinTheta = 1. / sinTheta;
-  const double cosPhi = x * invSinTheta;
-  const double sinPhi = y * invSinTheta;
   // prepare the jacobian to curvilinear
   FreeToBoundMatrix jacToCurv = FreeToBoundMatrix::Zero();
-  if (std::abs(cosTheta) < s_curvilinearProjTolerance) {
+  if (std::abs(z) < s_curvilinearProjTolerance) {
+    auto [cosPhi, sinPhi, cosTheta, sinTheta] =
+        VectorHelpers::evaluateTrigonomics(direction);
     // We normally operate in curvilinear coordinates defined as follows
-    jacToCurv(0, 0) = -sinPhi;
-    jacToCurv(0, 1) = cosPhi;
-    jacToCurv(1, 0) = -cosPhi * cosTheta;
-    jacToCurv(1, 1) = -sinPhi * cosTheta;
-    jacToCurv(1, 2) = sinTheta;
+    jacToCurv(eBoundLoc0, eFreePos0) = -sinPhi;
+    jacToCurv(eBoundLoc0, eFreePos1) = cosPhi;
+    jacToCurv(eBoundLoc1, eFreePos0) = -cosPhi * cosTheta;
+    jacToCurv(eBoundLoc1, eFreePos1) = -sinPhi * cosTheta;
+    jacToCurv(eBoundLoc1, eFreePos2) = sinTheta;
```

Comment:
Since we are modifying different positions, I would indicate, that this one doesn't change
```suggestion
    jacToCurv(eBoundLoc0, eFreePos0) = -sinPhi;
    jacToCurv(eBoundLoc0, eFreePos1) = cosPhi;
    // jacToCurv(eBoundLoc0, eFreePos2) = 0;
    jacToCurv(eBoundLoc1, eFreePos0) = -cosPhi * cosTheta;
    jacToCurv(eBoundLoc1, eFreePos1) = -sinPhi * cosTheta;
    jacToCurv(eBoundLoc1, eFreePos2) = sinTheta;
```